숫자야구 [STEP 1] Kiseok, Hisop

Hi-sop opened the pull request:
@leezion94  
Kiseok, Hisop 숫자야구 STEP1 PR입니다.

## 고민되었던 점
- 스트라이크와 볼 계산하는 함수 구현
    - 스트라이크가 아닐때만 볼을 검사
    - 볼을 검사하는 함수는 현재 number를 받아 검사하도록 함. 
    - 처음엔 스트라이크를 찾아내는 방법으로 Set의 교집합을 이용해보고자 했으나 볼을 찾아내기에 적합하지 않음. [Int]로 변경
  

- 들여쓰기 제한
    - 기존 이중 for문으로 이루어져있던 문장을 함수로 기능 분리하여 호출
  

- 함수와 변수의 네이밍
  

- 플로우차트 그리는 방법


## 조언을 얻고 싶은 부분
- 함수와 변수 네이밍이 적절했는지
- 함수별 기능 분리가 적절했는지

Commit:
Docs: README에 플로우차트 추가

carti1108 committed:
feat: step1 랜덤값 생성 함수, 비교 함수 추가:

Commit:
feat: 스트라이크 판별 기능 추가

carti1108 committed:
feat: 스트라이크와 볼 체크 함수 추가

Commit:
refactor: 출력 메세지 변경

Commit:
refactor: mainloop 함수 수정

실행 예시 구현

carti1108 committed:
refactor: 함수명 수정 및 승리조건 추가

mainLoop -> playNumberBaseBall로 변경, make -> generate로 변경
compareNumber함수 삭제

carti1108 committed:
refactor: 전달인자 위치 변경

Commit:
refactor: checkStrikeAndBall, checkBallForNumber 함수 수정

checkBallForNumber의 리턴값을 Bool타입으로 변경
checkStrikeAndBall의 for문 else분기를 else if로 변경하여 Ball함수의 리턴값을 바로 검사할수 있게 함
randomNumbers의 값은 변하지 않으므로 var -> let으로 변경

LeeZion94 commented:
안녕하세요 히솝, 기석 숫자 야구 리뷰를 맡게된 시온이라고 합니다. 잘부탁드립니다.

## 요구사항
----
현재 구현해주신 프로그램은 제시된 프로그램의 요구사항과는 일치하지 않는 부분이 존재합니다.
프로그램을 깔끔하고 가독성 좋게 구현해주신 부분은 칭찬드리고 싶지만 앞으로는 요구사항을 분명히 확인해주셨으면 좋겠습니다. 
```swift
남은 기회 : 0
```
부분은 요구사항에서는 출력하고 있지 않는 부분입니다. 확인부탁드립니다.

### 고민되었던 점
----
1. 스트라이크와 볼 계산하는 함수구현
- 스트라이크와 볼을 계산하는 함수를 구현할 수 있는 방법은 여러가지가 있을 수 있지만 최대한 가독성이 좋고 이해하기 쉽게 또는 최대한 반복문을 적게 돌면서 원하는 결과값을 도출해내는 것이 좋은 방법이라고 생각합니다. 이미 대부분의 코드에 있어서 잘작성해주셨지만 추가적으로 확인해보셨으면 하는 부분이 있어 코멘트로 남겨놓겠습니다.

2. 들여쓰기 제한
- 들여쓰기를 제한하는 연습을 해야하는 이유중 가장 큰 이유는 기능 분리를 습관화해야하기 때문입니다. 언급해주셨던 것 처럼 프로그램을 구현하기 위해서 어쩔 수 없이 2중 반복문이 나와야하는 경우라면 이번에 잘 구현해 주신대로 기능을 분리하여 구현하는 것이 좋은 습관으로 보입니다.

3. 함수와 변수의 네이밍
- 대부분의 경우에서 API Design GuideLine을 잘지켜주시면서 가독성 좋게 네이밍을 해주신 것 같습니다. 추가적으로 확인해주셨으면 하는 부분이 있어서 코멘트로 남겨두겠습니다.

4. FlowChart
- 우선 FlowChart의 경우에는 Step 1의 사용자 입력과 입력에 대한 검증이 포함된 것으로 보아 Step 1의 요구사항을 충족하고 잇는 FlowChart로 보이지는 않습니다. 이부분은 Step2가 나온 상황이기 때문에 의도하신 부분인지 궁금하네요.
- FlowChart는 말그대로 프로그램의 실행순서를 알기위해 그리는 Diagram입니다. 더 나아가서는 의사소토을 위한 도구와 자료가 될 수 있는 지표로써 활용되기도 합니다. 그런의미에서 FlowChart는 너무 복잡해서는 안된다고 생각합니다. 프로그램을 구현해보지 않은 사람도 FlowChart를 봤을 때 어느정도의 프로그램의 실행순서 및 흐름을 알아야한다는 것이 그 이유입니다. 

    그런 의미에 있어서 이번 FlowChart는 상징적은 표현을 사용해주어 잘작성해주신 것 같습니다.
### 조언을 얻고 싶은 부분
----
1. 함수와 변수 네이밍이 적절했는지
- API Design GuideLine에 맞게 잘 구현해주신것 같습니다. 비슷한 역할 및 목적을 가지는 변수 및 상수에도 통일성있게 네이밍을 해주셔서 새롭게 코드를 보는 제3자의 입장에서 이해하기 훨씬 쉬웠습니다.
- 함수의 네이밍 또한 동사로 시작하면서 함수명만 확인하더라도 이해하기 쉽게 잘 구현되어서 좋았습니다.

2. 함수별 기능 분리가 적절했는지
- 의도해주시는 함수명에 맞게 역할이 잘 분배되었다고 생각합니다. 또한 각각의 함수들이 더이상 길어지지 않게 기능분리도 잘되어있다고 생각합니다. 기능분리에 관한 추가적인 피드백은 코멘트로 남겨두겠습니다. 저는 한 함수가 최대한 20줄을 넘지 않게 구현하려고 노력하는 편입니다. 이는 'Clean Code'에도 나와있는 기준이기도 합니다. 꼭 20줄이 아니여도 되겠지만 과제를 구현하면서 많이 노력해주신 것 처럼 앞으로도 일정한 줄 수에 대한 기준이나 분리를 해야하는 역할에 대한 기준으로 가지고 코드를 구현하신다면 좋은 습관을 가지실 수 있을 것 같습니다.

## 스텝진행
- 이후 스텝의 진행은 해당 Step 1이 머지가 되어야 진행하실 수 있습니다. 제가 남겨드린 코멘트와 질문들을 확인하시고 꼭 답변을 남겨주세요. (PR, Comment 답변을 기다리실 때 새로운 브랜치에 이후 작업을 진행하시는 것은 무관합니다.)

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,58 @@
 
 import Foundation
```
현재 프로그램에는 필요가 없는 문장으로 보입니다.
Foundation은 어떤 것이고, 왜 Import 하셨나요?

Hi-sop replied:
https://developer.apple.com/documentation/foundation
저희는 Foundation이 기본적인 타입들을 사용하는데 있어서 필수적인 프레임워크라고 생각했습니다.

https://developer.apple.com/documentation/swift/swift-standard-library
찾아보니 기본적인 타입들은 스탠다드 라이브러리에 이미 포함되어 있었습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,58 @@
 
-print("Hello, World!")
+func playNumberBaseBallGame() {
+    let randomNumbers: [Int] = generateRandomNumber()
```
Random한 난수를 생성받아 프로그램을 진행하는 것은 맞지만, 컴퓨터와 유저의 숫자야구 게임입니다.
생성될 때는 난수이지만 할당될 때에는 Computer가 가지는 수가 아닐까요?

Hi-sop replied:
ComputerRandomNumbers가 더 적절할 것 같네요. 다음 커밋에 반영하겠습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,58 @@
+        numbers.insert(Int.random(in: 1...9))
+    }
+    
+    return numbers.map { Int($0) }
```
고차함수를 사용하신 것으로 보입니다.
고차함수는 어떤 함수들을 고차함수라고 하나요?
또한, 사용하신 map은 어떤 동작을 하고 있나요?

carti1108 replied:
고차함수는 다른 함수를 전달인자로 받거나 함수실행의 결과를 함수로 반환하는 함수입니다.
map은 주어진 컨테이너의 요소에 작성한 클로저를 맵핑하여 반환하는 동작을 하고있습니다.
저희가 구현한 코드에서는 Set<Int>를 Int로 매핑하려고 하였습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,58 @@
+        if strike == 3 {
+            print("사용자 승리!")
+        }
+        chance -= 1
```
구현에 주신 프로그램의 컨벤션과 맞지 않는 부분으로 보입니다.
제가 이해한 컨벤션은 조건문 이후의 개행 입니다. 의도하신 부분인가요?

Hi-sop replied:
낮은 확률이지만 숫자가 모두 일치해 플레이어가 승리하는 분기를 만들었습니다.
return은 빼먹은 것 같습니다. 다음 커밋에 추가하도록 하겠습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,58 @@
+    while chance != 0 {
+        userRandomNumbers = generateRandomNumber()
+        print("임의의 수 : \(userRandomNumbers.map { String($0) }.joined(separator: " "))")
+        let (strike, ball) = checkStrikeAndBall(userNumbers: userRandomNumbers, randomNumbers: randomNumbers)
```
checkStrikeAndBall 이라는 함수는 User와 Computer의 숫자를 비교하여 Strike, Ball의 결과를 도출하는 함수로 보입니다.
userNumber는 눈에 보이지만 computerNumbers는 한눈에 확인하기 어렵다고 보입니다. 의도하신 부분인가요?

carti1108 replied:
제시받은 step1의 실행예시를 구현하는 것을 목표로 하였고, 이후 step의 실행예시에서도 computerNumbers의 값이 보여지지 않아 구현할 필요가 없다고 생각하였습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,58 @@
+}
 
+func generateRandomNumber() -> [Int] {
+    var numbers: Set<Int> = []
```
Set을 활용하신 부분 너무 좋습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,58 @@
+}
+
+func checkBallForNumber(userNumber: Int, randomNumbers: [Int]) -> Bool {
+    for randomNumber in randomNumbers {
```
배열에 해당 숫자가 포함된 것을 확인하고 있는 로직으로 보입니다. 반복문을 직접구현하여 확인하는 것도 좋지만 이미 구현된 로직을 사용하는 것도 가독성에 있어서 좋은 방법으로 보입니다. 아래 링크를 확인해주세요!

https://developer.apple.com/documentation/swift/array/contains(_:)

Hi-sop replied:
추가 수정을 할까 고민했던 부분입니다! 다음 커밋에 반영해볼게요.
checkBallForNumber 함수는 삭제해도 될 것 같아요

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,58 @@
+    for (index, number) in userNumbers.enumerated() {
+        if number == randomNumbers[index] {
+            strike += 1
+        } else if checkBallForNumber(userNumber: number, randomNumbers: randomNumbers) {
```
else인 경우를 고려하지 않는 이유는 이해하지만 처리하지 않는 case가 있다면 else if의 사용이 코드를 처음보는 사람의 입장에서는 바로바로 이해하지 않는 부분이 될 수 있다고 생각합니다. 
else로 두고 내부에 처리하는 것이 가독성이 더 좋을 수 있다고 생각하는데 이 부분에 대해서는 어떻게 생각하시나요?

carti1108 replied:
협의해본 결과 else 만 사용하는 것이 가독성 부분에서는 더 나을것 같았으나 내부에서 처리하려면 들여쓰기가 3번이상 쓰이기 때문에 else if를 사용하는 것을 유지하기로 결정하였습니다.

LeeZion94 commented on NumberBaseball/NumberBaseball/main.swift:
```diff
@@ -6,5 +6,58 @@
+        userRandomNumbers = generateRandomNumber()
+        print("임의의 수 : \(userRandomNumbers.map { String($0) }.joined(separator: " "))")
+        let (strike, ball) = checkStrikeAndBall(userNumbers: userRandomNumbers, randomNumbers: randomNumbers)
+        print("\(strike) 스트라이크, \(ball) 볼")
```
변수 및 상수 다음줄은 개행을 하는 컨벤션을 가지고 있는 것으로 보입니다.
해당 print는 그렇지 않습니다만 의도하신 부분인가요?

Hi-sop replied:
놓친 부분입니다. 수정하도록 하겠습니다!

carti1108 committed:
refactor: checkballfornumber함수를 contains로 변경후 삭제

Hi-sop committed:
STEP1 기준으로 플로우차트 수정

carti1108 committed:
refactor: 코멘트 받은 사항 수정

스트라이크가 3일 때 함수가 종료될 수 있도록 수정
남은 기회가 0일 때 출력하지 않도록 변경
import Foundation 제거
randomNumber 상수명을 computerRandomNumber로 변경
let 변수 선언 후 개행하는 컨벤션으로 정리
